### PR TITLE
ASRC-10: support uploading new procedure set

### PR DESCRIPTION
**Pull request by dgtombs**

### CODE REVIEW

File: srcalc/src/main/java/gov/va/med/srcalc/service/DefaultAdminService.java [hunk 1]
```diff
@@ -107,4 +114,26 @@ public void saveVariable(final AbstractVariable variable)
...
+    {
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        final int deleteCount = fProcedureDao.replaceAllProcedures(newProcedures);
+        stopwatch.stop();
+        
+        fLogger.info(
+                "Replaced all {} Procedures in the DB with a new set of {} in {}ms.",
+                deleteCount,
+                newProcedures.size(),
+                stopwatch.elapsed(TimeUnit.MILLISECONDS));
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/service/DefaultAdminService.java]: I like logging the time metric here.

File: srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    
+    private final int fRowNumber;
+    private final String fColumnName;
+    private final String[] fCodes;
+    private final Object[] fArguments;
+    private final String fDefaultMessage;
+    
+    /**
+     * Constructs an instance.
+     * @param rowNumber the associated row's number (1-based). May also be {@link #ROW_NUMBER_GLOBAL}
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java]: I'm assuming this is 1-based because it's more understandable to a user.

File: srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    public TabularUploadError(
+            final int rowNumber,
+            final String columnName,
+            final String[] codes,
+            final Object[] arguments,
+            final String defaultMessage)
+    {
+        fRowNumber = rowNumber;
+        fColumnName = Objects.requireNonNull(columnName, "field must not be null");
+        fCodes = codes.clone();
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java]: Are we sure that a shallow copy is alright here?

File: srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    }
+    
+    /**
+     * Creates an error for the whole upload.
+     * @param code the error code. Must be non-null.
+     * @param arguments any necessary message arguments. May be null.
+     * @param defaultMessage the default message to use if no codes could be resolved
+     * @return a new instance
+     */
+    public static TabularUploadError global(
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java]: I think it would help to describe what a global error is here. Maybe one or two examples would help.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/EditProceduresController.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            return new ModelAndView("redirect:" + BASE_URL);
+        }
+        // Otherwise, display the errors to the user.
+        else
+        {
+            fLogger.debug("There were errors: {}", result.getErrors());
+            final ModelAndView mav = displayCurrentProcedures();
+            mav.addObject("validationErrors", result.getErrors());
+            return mav;
+        }
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/EditProceduresController.java]: Not a huge complaint, but this is the opposite ordering that we have done other controllers in when referring to errors (the test for errors have usually come first).

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+     * @param rowNumber the row number for any validation errors (1-based)
+     * @return the translated Procedure, if there were no validation errors
+     */
+    private Optional<Procedure> translateRecord(
+            final CSVRecord record,
+            final Collection<TabularUploadError> errors,
+            final int rowNumber)
+    {
+        // Construct our own error list to detect if we added any errors.
+        final ArrayList<TabularUploadError> ourErrors = new ArrayList<>();
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: I feel that something like uploadErrors would be a more descriptive name.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        
+        final String shortDesc = tryGetValue(record, 5);
+        ValidationUtils2.validateRequiredString(
+                shortDesc, rowNumber, "shortDescription", Procedure.DESCRIPTION_MAX, ourErrors);
+        
+        final String rvuString = tryGetValue(record, 1);
+        float rvu = Float.NaN;
+        try
+        {
+            rvu = Float.parseFloat(rvuString);
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: There is a Floats.tryParse() method from Guava that doesn't require an exception to be caught.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+            ourErrors.add(TabularUploadError.forField(
+                    rowNumber,
+                    "eligible",
+                    boolean.class,
+                    ValidationCodes.TYPE_MISMATCH,
+                    null,
+                    ex.getMessage()));
+        }
+            
+        // Only create the procedure if we did not record any validation ourErrors for
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: This sentence sounds weird, like you meant to put "validation errors" instead of what is there.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+     * @param records must be non-empty
+     * @param errors for recording validation errors
+     */
+    private ParseResult readProcedures(
+            final List<CSVRecord> records, final Collection<TabularUploadError> errors)
+    {
+        int firstRow = 0;
+        
+        // Detect if the first row is a header and strip if so.
+        if (records.get(0).get(0).equalsIgnoreCase("cpt"))
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: Nothing wrong with this code, but it should be stated somewhere in documentation about what the header would look like if it were included.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        finally
+        {
+            try
+            {
+                csvReader.close();
+            }
+            catch (final IOException ex)
+            {
+                fLogger.warn("Filed to close the CSV reader.", ex);
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: Should be "Failed to close..."

File: srcalc/src/main/resources/messages.properties [hunk 11]
```diff
@@ -5,10 +5,15 @@ tooLowExclusive = Value must be greater than {0}.
 tooHighInclusive = Value must be less than or equal to {0}.
 tooHighExclusive = Value must be less than {0}.
 typeMismatch.float = Please enter a valid number.
+typeMismatch.cptCode = Please enter a 5-character CPT code.
 noSelection = Please select an option.
 tooLong.java.lang.String = Please enter {0} characters or less.
+# This is used when a String is fixed-length and the given String didn't match. (It could
+# have been too short or too long.)
+badFixedLength.java.lang.String = Value must be {0} characters.
```
Comment (jamesbigmac) [inline srcalc/src/main/resources/messages.properties]: Do you think we should make an error message specific to CPT codes?

File: srcalc/src/main/webapp/WEB-INF/web.xml [hunk 12]
```diff
@@ -21,6 +21,10 @@
 			<param-value></param-value>
 		</init-param>
 		<load-on-startup>1</load-on-startup>
+        <multipart-config>
+            <!-- 10 million bytes -->
+            <max-file-size>10000000</max-file-size>
```
Comment (jamesbigmac) [inline srcalc/src/main/webapp/WEB-INF/web.xml]: This should be specified in documentation outside the code as well.

File: srcalc/src/main/webapp/WEB-INF/views/editProcedures.jsp [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    </c:if>
+    <c:if test="${!empty validationErrors}">
+    <div class="error">
+    <p>The uploaded procedure set was not valid. The database has not been modified.</p>
+    <p>Errors were:</p>
+    <ul>
+        <c:forEach var="error" items="${validationErrors}">
+        <li>${error.locationPrefix}<spring:message message="${error}" /></li>
+        </c:forEach>
+    </ul>
```
Comment (jamesbigmac) [inline srcalc/src/main/webapp/WEB-INF/views/editProcedures.jsp]: I feel like the errors should be after the upload button so that the button is easier to find than having to scroll to the bottom of the page.

File: srcalc/src/test/java/gov/va/med/srcalc/util/TabularUploadErrorTest.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        assertEquals(code, error.getCode());
+        assertNull(null, error.getArguments());
+        assertEquals(defaultMessage, error.getDefaultMessage());
+        
+        assertEquals("", error.getLocationPrefix());
+        
+        
+        // The class provides no documentation on what toString() will contain, but at
+        // least make sure it is non-empty (and doesn't throw an Exception).
+        assertNotEquals("", error.toString());
```
Comment (jamesbigmac) [inline srcalc/src/test/java/gov/va/med/srcalc/util/TabularUploadErrorTest.java]: Should this also test to see if the toString method doesn't return null?

File: srcalc/src/test/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReaderTest.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        /* Verification */
+        assertEquals(0, result.getErrors().size());
+        assertEquals(VALID_PROCEDURES, result.getProcedures());
+    }
+    
+    @Test
+    public final void testReadCorruptCsv() throws IOException
+    {
+        final CharSource corruptCsv = Resources.asCharSource(
+                Resources.getResource(getClass(), "invalid.csv"), Charsets.US_ASCII);
```
Comment (jamesbigmac) [inline srcalc/src/test/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReaderTest.java]: The URL could also be a constant in this class.

File: srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java (hunk 2 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/util/TabularUploadError.java]: I'll update the Javadocs to state the rationale.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/EditProceduresController.java (hunk 5 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/EditProceduresController.java]: True, consistency is good, but I think it's better this way than negating the if condition.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java (hunk 7 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: Hmm, I like that method, but it's marked Beta. I think I'll avoid it for now.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java (hunk 9 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/admin/ProcedureCsvReader.java]: Good comment. I meant to provide a sample upload and you reminded me.

File: srcalc/src/main/resources/messages.properties (hunk 11 above)
Comment (dgtombs) [inline srcalc/src/main/resources/messages.properties]: I think this looks fine on the UI. Do you disagree?

File: srcalc/src/main/webapp/WEB-INF/web.xml (hunk 12 above)
Comment (dgtombs) [inline srcalc/src/main/webapp/WEB-INF/web.xml]: Where? In the UI? Or just the technical manual somewhere?
